logictest: TestLogic_prepare with high vmodule failed

pkg/sql/logictest/tests/local/local_test.TestLogic_prepare [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_SqlLogicTestHighVModuleNightlyBazel/11245460?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_SqlLogicTestHighVModuleNightlyBazel/11245460?buildTab=artifacts#/) on master @ [a3a389e464defd3d301a0b080435cf33502a02d2](https://github.com/cockroachdb/cockroach/commits/a3a389e464defd3d301a0b080435cf33502a02d2):


```
*   | github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*Server).serveImpl.func3
*   | 	github.com/cockroachdb/cockroach/pkg/sql/pgwire/server.go:997
*   | runtime.goexit
*   | 	GOROOT/src/runtime/asm_amd64.s:1594
* Wraps: (4) runtime error: invalid memory address or nil pointer dereference
* Error types: (1) *withstack.withStack (2) *safedetails.withSafeDetails (3) *withstack.withStack (4) runtime.errorString
*
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x38 pc=0x36e1486]

goroutine 40399 [running]:
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).closeWrapper(0xc005721980, {0x6bb0f20, 0xc0051e4a50}, {0x4e2c5c0?, 0x995f240?})
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1199 +0x1b4
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1({0x6bb0f20, 0xc0051e4a50}, {0x3d1c7471?})
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:952 +0x4a
panic({0x4e2c5c0, 0x995f240})
	GOROOT/src/runtime/panic.go:884 +0x212
github.com/cockroachdb/cockroach/pkg/sql.(*planner).getCommonSQLEventDetails(0xc005721fc8, {0x0})
	github.com/cockroachdb/cockroach/pkg/sql/event_log.go:199 +0x1c6
github.com/cockroachdb/cockroach/pkg/sql.(*planner).logEventsWithOptions(0xc005721fc8, {0x6bb0f20, 0xc0068b3710}, 0x1, {0x1?, 0x9a5ac88?, {0x0?}}, {0xc0011c4430, 0x1, 0x1})
	github.com/cockroachdb/cockroach/pkg/sql/event_log.go:220 +0x8b
github.com/cockroachdb/cockroach/pkg/sql.(*planner).maybeLogStatementInternal(0xc005721fc8, {0x6bb0f20, 0xc0068b3710}, 0x0, 0x0, 0x16, 0x0, 0x1, 0x0, {0x6b6ee00, ...}, ...)
	github.com/cockroachdb/cockroach/pkg/sql/exec_log.go:260 +0x9d7
github.com/cockroachdb/cockroach/pkg/sql.(*planner).maybeLogStatement(0x0?, {0x6bb0f20, 0xc0068b3710}, 0xc0044daa20?, 0x2?, 0xc005721980?, 0xc0011c4400?, 0xc0047ad1a0?, 0x6bdb6a0?, {0x6b6ee00, ...}, ...)
	github.com/cockroachdb/cockroach/pkg/sql/exec_log.go:134 +0x14e
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState.func15()
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:803 +0x366
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState(0xc005721980, {0x6bb0f20, 0xc0068b3710}, {{0x6bdb6a0, 0xc000fa24c0}, {0x0, 0x0, 0x0}, {0xc00626817b, 0x16}, ...}, ...)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:909 +0x584f
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt.func1({0x6bb0f20?, 0xc0068b34d0?})
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:144 +0xbc
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithProfiling(0xc006f53260?, {0x6bb0f20?, 0xc0068b34d0?}, {0x6bdb6a0?, 0xc000fa24c0?}, 0x100000000?, 0x1?)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:3118 +0x330
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt(0xc005721980, {0x6bb0f20, 0xc0068b34d0}, {{0x6bdb6a0, 0xc000fa24c0}, {0x0, 0x0, 0x0}, {0xc00626817b, 0x16}, ...}, ...)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:143 +0x55e
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd.func1(0xc006f53988, 0xc005721980, 0xc006f534e0, 0xc006f535b0, 0xc006f535f0, 0xc006f535d0, 0xc006f535c0, 0xc006f535e0)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:2220 +0x319
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd(0xc005721980)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:2225 +0x1167
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run(0xc005721980, {0x6bb0f20, 0xc0051e4a50}, 0x4121187e00000000?, 0xc005b75cc8?, 0xc00747e520?)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:2142 +0x216
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn(0xc007392c80?, {0x6bb0f20?, 0xc0051e4a50?}, {0xc006474046?}, 0x4?, 0xc0051e4570?)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:954 +0xe6
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommands(0xc004a21800, {0x6bb0f20, 0xc0051e4a50}, {0x0, 0x4, {0x1, {0x561c26c, 0x3}, {0xc003f0a310, 0xf}, ...}, ...}, ...)
	github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:247 +0x405
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*Server).serveImpl.func3()
	github.com/cockroachdb/cockroach/pkg/sql/pgwire/server.go:997 +0xf5
created by github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*Server).serveImpl
	github.com/cockroachdb/cockroach/pkg/sql/pgwire/server.go:994 +0x688
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-queries
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestLogic_prepare.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-30460

--- a/pkg/sql/conn_executor_exec.go
+++ b/pkg/sql/conn_executor_exec.go
@@ -903,16 +903,18 @@ func (ex *connExecutor) execStmtInOpenState(
 		// PREPARE statement itself.
 		oldPlaceholders := p.extendedEvalCtx.Placeholders
 		p.extendedEvalCtx.Placeholders = nil
+		defer func() {
+			// The call to addPreparedStmt changed the planner stmt to the
+			// statement being prepared. Set it back to the PREPARE statement,
+			// so that it's logged correctly.
+			p.stmt = stmt
+			p.extendedEvalCtx.Placeholders = oldPlaceholders
+		}()
 		if _, err := ex.addPreparedStmt(
 			ctx, name, prepStmt, typeHints, rawTypeHints, PreparedStatementOriginSQL,
 		); err != nil {
 			return makeErrEvent(err)
 		}
-		// The call to addPreparedStmt changed the planner stmt to the statement
-		// being prepared. Set it back to the PREPARE statement, so that it's
-		// logged correctly.
-		p.stmt = stmt
-		p.extendedEvalCtx.Placeholders = oldPlaceholders
 		return nil, nil, nil
 	}
 

